server/src/infra/repositories/media.repository.ts

```diff
@@ -124,4 +124,31 @@ export class MediaRepository implements IMediaRepository {
         .run();
     });
   }
+
+  async generateThumbhash(imagePath: string): Promise<Buffer> {
+    const maxSize = 100;
+
+    const { data, info } = await sharp(imagePath)
+      .resize(maxSize, maxSize, { fit: 'inside', withoutEnlargement: true })
+      .raw()
+      .ensureAlpha()
+      .toBuffer({ resolveWithObject: true });
+
+    // this is a hackey workaround to get the official thumbhash library to work on CommonJS
+    // shouldn't cause any issues, but it is a tad jank
+    return new Promise<Buffer>((resolve, reject) => {
```

Comment:
Usage of the official module is awesome, thank you, though I don't think this hack is really required. I think it's possible to use `import` directly, such:

```ts
return new Promise<Buffer>((resolve, reject) => import('thumbhash')
    .then(thumbhash => resolve(Buffer.from(thumbhash.rgbaToThumbHash(info.width, info.height, data))))
    .catch(reject));
```

Reply:
Or maybe?

```ts
const thumbhash = await import('thumbhash');
return Buffer.from(thumbhash.rgbaToThumbHash(info.width, info.height, data));
```

Then change the return type from `Promise<Buffer>` to `Buffer`. Not sure if there's any advantage to using a promise otherwise.

Reply:
Though I am confused as someone seems to suggest that they are able to use `require` with thumbhash.

https://github.com/evanw/thumbhash/issues/2#issuecomment-1481848612

Reply:
fwiw I just tested this diff and it works.

```diff
diff --git a/server/src/infra/repositories/media.repository.ts b/server/src/infra/repositories/media.repository.ts
index 4359e741..6e43bab7 100644
--- a/server/src/infra/repositories/media.repository.ts
+++ b/server/src/infra/repositories/media.repository.ts
@@ -4,6 +4,8 @@ import ffmpeg, { FfprobeData } from 'fluent-ffmpeg';
 import sharp from 'sharp';
 import { promisify } from 'util';
 import fs from 'fs/promises';
+import { rgbaToThumbHash } from 'thumbhash';
+
 
 const probe = promisify<string, FfprobeData>(ffmpeg.ffprobe);
 
@@ -133,22 +135,6 @@ export class MediaRepository implements IMediaRepository {
       .raw()
       .ensureAlpha()
       .toBuffer({ resolveWithObject: true });
-
-    // this is a hackey workaround to get the official thumbhash library to work on CommonJS
-    // shouldn't cause any issues, but it is a tad jank
-    return new Promise<Buffer>((resolve, reject) => {
-      (eval('import("thumbhash")') as Promise<typeof import('thumbhash')>)
-        .then((thumbhash_lib) => {
-          try {
-            const thumbhash = thumbhash_lib.rgbaToThumbHash(info.width, info.height, data);
-            resolve(Buffer.from(thumbhash));
-          } catch (error) {
-            reject(error);
-          }
-        })
-        .catch((error) => {
-          reject(error);
-        });
-    });
+    return rgbaToThumbHash(info.width, info.height, data);
   }
 }

```

Reply:
> Or maybe?
> 
> ```ts
> const thumbhash = await import('thumbhash');
> return Buffer.from(thumbhash.rgbaToThumbHash(info.width, info.height, data));
> ```
> 
> The return type doesn't need to change as async functions always return a promise.

Nope :(
```ts 
 [Nest] 29  - 06/14/2023, 11:44:48 PM   ERROR [JobService] Unable to run job handler: Error [ERR_REQUIRE_ESM]: require() of ES Module /usr/src/app/node_modules/thumbhash/thumbhash.js from /usr/src/app/dist/infra/repositories/media.repository.js not supported.```

Reply:
> Usage of the official module is awesome, thank you, though I don't think this hack is really required. I think it's possible to use `import` directly, such:
> 
> ```ts
> return new Promise<Buffer>((resolve, reject) => import('thumbhash')
>     .then(thumbhash => resolve(Buffer.from(thumbhash.rgbaToThumbHash(info.width, info.height, data))))
>     .catch(reject));
> ```

Also doesn't work 
```ts
Instead change the require of thumbhash.js in /usr/src/app/dist/infra/repositories/media.repository.js to a dynamic import() which is available in all CommonJS modules.
immich_microservices     | [Nest] 168  - 06/14/2023, 11:47:59 PM   ERROR [JobService] Error [ERR_REQUIRE_ESM]: require() of ES Module /usr/src/app/node_modules/thumbhash/thumbhash.js from /usr/src/app/dist/infra/repositories/media.repository.js not supported.
immich_microservices     | Instead change the require of thumbhash.js in /usr/src/app/dist/infra/repositories/media.repository.js to a dynamic import() which is available in all CommonJS modules.
immich_microservices     |     at /usr/src/app/dist/infra/repositories/media.repository.js:152:91
immich_microservices     | [Nest] 168  - 06/14/2023, 11:47:59 PM   ERROR [JobService] Object:
immich_microservices     | {
immich_microservices     |   "id": "7fca3908-fe98-4525-99f1-1137c84b2306"
immich_microservices     | }```
I'm a little at a loss at how to make this work right.

Reply:
You tried

```ts
import { rgbaToThumbHash } from 'thumbhash';
```
and

```ts
return rgbaToThumbHash(info.width, info.height, data);
```
?

Reply:
> fwiw I just tested this diff and it works.
> 
> ```diff
> diff --git a/server/src/infra/repositories/media.repository.ts b/server/src/infra/repositories/media.repository.ts
> index 4359e741..6e43bab7 100644
> --- a/server/src/infra/repositories/media.repository.ts
> +++ b/server/src/infra/repositories/media.repository.ts
> @@ -4,6 +4,8 @@ import ffmpeg, { FfprobeData } from 'fluent-ffmpeg';
>  import sharp from 'sharp';
>  import { promisify } from 'util';
>  import fs from 'fs/promises';
> +import { rgbaToThumbHash } from 'thumbhash';
> +
>  
>  const probe = promisify<string, FfprobeData>(ffmpeg.ffprobe);
>  
> @@ -133,22 +135,6 @@ export class MediaRepository implements IMediaRepository {
>        .raw()
>        .ensureAlpha()
>        .toBuffer({ resolveWithObject: true });
> -
> -    // this is a hackey workaround to get the official thumbhash library to work on CommonJS
> -    // shouldn't cause any issues, but it is a tad jank
> -    return new Promise<Buffer>((resolve, reject) => {
> -      (eval('import("thumbhash")') as Promise<typeof import('thumbhash')>)
> -        .then((thumbhash_lib) => {
> -          try {
> -            const thumbhash = thumbhash_lib.rgbaToThumbHash(info.width, info.height, data);
> -            resolve(Buffer.from(thumbhash));
> -          } catch (error) {
> -            reject(error);
> -          }
> -        })
> -        .catch((error) => {
> -          reject(error);
> -        });
> -    });
> +    return rgbaToThumbHash(info.width, info.height, data);
>    }
>  }
> ```

I'm not sure what to say, this just doesn't work. 

```ts
/usr/src/app/dist/infra/repositories/media.repository.js:12
immich_server            | const thumbhash_1 = require("thumbhash");
immich_server            |                     ^
immich_server            | Error [ERR_REQUIRE_ESM]: require() of ES Module /usr/src/app/node_modules/thumbhash/thumbhash.js from /usr/src/app/dist/infra/repositories/media.repository.js not supported.
immich_server            | Instead change the require of thumbhash.js in /usr/src/app/dist/infra/repositories/media.repository.js to a dynamic import() which is available in all CommonJS modules.```

Reply:
I'm confused, what are you running?

`npm run test` works okay, is there something strange going on with bundling?

Reply:
> I'm confused, what are you running?
> 
> `npm run test` works okay, is there something strange going on with bundling?

Jest doesn't test this well (I'm not sure why). You have to try to actually generate thumbnails with the full stack up to see the errors.

Reply:
Could you try changing this line from "es2017" to "es2018"?

https://github.com/immich-app/immich/blob/408fa45c51822704510c9741a027579ef0cc4ecc/server/tsconfig.json#L11